extensions/wikia/UserLogin/tests/UserLoginTest.php

```diff
@@ -1,782 +1,869 @@
+					'returnto' => 'Special:UserLogin',
+					'editToken' => '123456789',
 				],
-				[
-					'query' => [
-						'login' => self::TEST_USERNAME,
-						'returnto' => 'Special:UserLogin',
-						// notice the returnto&login in previous returntoquery were different
-						// and are supposed to be overwritten
-						'returntoquery' => 'returnto=Main_page%3Alogin=Test%3AeditToken=123456789',
-						'editToken' => '123456789',
-					],
-					'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'wfCgiToArrayResult' => [],
-					'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'message' => 'Query with returntoquery',
+				'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'wfCgiToArrayResult' => [],
+				'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'message' => 'Query without returntoquery parameter',
+			],
+			[
+				'query' => [
+					'login' => self::TEST_USERNAME,
+					'returnto' => 'Special:UserLogin',
+					// notice the returnto&login in previous returntoquery were different
+					// and are supposed to be overwritten
+					'returntoquery' => 'returnto=Main_page%3Alogin=Test%3AeditToken=123456789',
+					'editToken' => '123456789',
 				],
-			];
-		}
+				'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'wfCgiToArrayResult' => [],
+				'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'message' => 'Query with returntoquery',
+			],
+		];
+	}
+
+	/**
+	 * @param string $page
+	 * @param string $queryString
+	 * @param string $extraQueryString
+	 * @param int $cbVal
+	 * @param Title $actualTitle
+	 * @param string $actualQueryString
+	 *
+	 * @dataProvider getRedirectUrlDataProvider
+	 */
+	public function testGetRedirectUrl( $page, $queryString, $extraQueryString, $cbVal, Title $actualTitle, $actualQueryString ) {
+		$request = F::app()->wg->Request;
+		$request->setVal( 'returnto', $page );
+		$request->setVal( 'returntoquery', $queryString );
+
+		$userLoginHelper = new UserLoginHelper();
+
+		$testUrl = $userLoginHelper->getRedirectUrl( $extraQueryString, $cbVal );
+		$actualUrl = $actualTitle->getFullUrl( $actualQueryString );
+
+		$this->assertEquals( $testUrl, $actualUrl );
+	}
 
+	public function getRedirectUrlDataProvider() {
+		return [
+			// test title w/o query string
+			[
+				'Foo',
+				'',
+				'fbconnected=1',
+				123,
+				Title::newFromText( 'Foo' ),
+				'fbconnected=1&cb=123'
+			],
+
+			// test title w/ query string
+			[
+				'Foo',
+				'foo=bar',
+				'fbconnected=1',
+				456,
+				Title::newFromText( 'Foo' ),
+				'foo=bar&fbconnected=1&cb=456'
+			],
+
+			// test forbidden return-to page w/o query string
+			[
+				'Special:Signup',
```

Comment:
'Special:Signup' redirects to 'Special:UserLogin'. Should it be changed everywhere or is this a NS defined somewhere?
Also is there a case for `:UserSignup`?